File: panghae/PGS/알고리즘 고득점kit/깊이-너비 우선 탐색(DFS-BFS)/게임 맵 최단거리.py
from collections import deque


def solution(maps):
    answer = -1
    q = deque()

    q.append((0, 0, 1))

    dx = [-1, 0, 1, 0]
    dy = [0, 1, 0, -1]

    while q:
        x, y, count = q.popleft()

        if x == len(maps) -1 and y == len(maps[0]) - 1:
            answer = count
            break

        if maps[x][y] == 1:
            maps[x][y] = 0

            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < len(maps) and 0 <= ny < len(maps[0]) and maps[nx][ny] == 1:
                    q.append((nx, ny, count + 1))

    return answer


print(solution([[1,0,1,1,1],[1,0,1,0,1],[1,0,1,1,1],[1,1,1,0,1],[0,0,0,0,1]]))
print(solution([[1,0,1,1,1],[1,0,1,0,1],[1,0,1,1,1],[1,1,1,0,0],[0,0,0,0,1]]))

File: panghae/PGS/알고리즘 고득점kit/깊이-너비 우선 탐색(DFS-BFS)/게임 맵 최단거리.js
function solution(maps) {
    let answer = -1;

    const xLength = maps.length;
    const yLength = maps[0].length;

    const dx = [-1, 0, 1, 0];
    const dy = [0, 1, 0, -1];

    let q = [[[0, 0], 1]];

    while (q.length) {
        const [[x, y], count] = q.shift();

        for (let i = 0; i < 4; i += 1) {
            const nx = x + dx[i];
            const ny = y + dy[i];
            if (0 <= nx && nx < xLength && 0 <= ny && ny < yLength && maps[nx][ny] === 1) {
                if (nx === xLength - 1 && ny === yLength - 1) {
                    answer = count + 1;
                    q = [];
                    break;
                }
                q.push([[nx, ny], count + 1]);
                maps[nx][ny] = 0;
            }
        }
    }

    return answer;
}

console.log(solution([[1, 0, 1, 1, 1], [1, 0, 1, 0, 1], [1, 0, 1, 1, 1], [1, 1, 1, 0, 1], [0, 0, 0, 0, 1]]));
console.log(solution([[1, 0, 1, 1, 1], [1, 0, 1, 0, 1], [1, 0, 1, 1, 1], [1, 1, 1, 0, 0], [0, 0, 0, 0, 1]]));